tools/clang/unittests/HLSLExec/ExecutionTest.cpp

```diff
@@ -13236,6 +13205,173 @@ void ExecutionTest::WaveSizeTest() {
+           prefShaderWaveSize <= maxShaderWaveSize; prefShaderWaveSize *= 2) {
+
+        // For now, only allow valid shader wave ranges
+        // TODO: Create callback function for CreateComputePipelineState to
+        // handle expected errors.
+        bool AcceptedByRuntime = TestShaderRangeAgainstRequirements(
+            minShaderWaveSize, maxShaderWaveSize, minWaveSize, maxWaveSize);
+        if (!AcceptedByRuntime) {
+          LogCommentFmt(
+              L"Wave size range [%d, %d, %d] not supported by runtime, "
+              L"which only supports range [%d, %d]",
+              minShaderWaveSize, maxShaderWaveSize, prefShaderWaveSize,
+              minWaveSize, maxWaveSize);
+          continue;
+        }
+
+        // format compiler args
+        char compilerOptions[64];
+        VERIFY_IS_TRUE(
+            sprintf_s(compilerOptions, sizeof(compilerOptions),
+                      "-D MINWAVESIZE=%d -D MAXWAVESIZE=%d -D PREFWAVESIZE=%d",
+                      minShaderWaveSize, maxShaderWaveSize,
+                      prefShaderWaveSize) != -1);
+
+        // run the shader
+        std::shared_ptr<ShaderOpTestResult> test = RunShaderOpTestAfterParse(
+            pDevice, m_support, "WaveSizeRangeTest",
+            [&](LPCSTR Name, std::vector<BYTE> &Data, st::ShaderOp *pShaderOp) {
+              VERIFY_IS_TRUE((0 == strncmp(Name, "UAVBuffer0", 10)));
+              pShaderOp->Shaders.at(0).Arguments = compilerOptions;
+              pShaderOp->Shaders.at(0).Text = waveSizeTestShader;
+
+              VERIFY_IS_TRUE(sizeof(WaveSizeTestData) * MAX_WAVESIZE <=
+                             Data.size());
+              WaveSizeTestData *pInData = (WaveSizeTestData *)Data.data();
+              memset(pInData, 0, sizeof(WaveSizeTestData) * MAX_WAVESIZE);
+            },
+            ShaderOpSet);
+
+        // verify expected values
+        MappedData dataUav;
+        WaveSizeTestData *pOutData;
+
+        LogCommentFmt(
+            L"Verifying test result for these shader wave size ranges:");
+        LogCommentFmt(L"Minimum shader wave size: %d", minShaderWaveSize);
+        LogCommentFmt(L"Maximum shader wave size: %d", maxShaderWaveSize);
+        LogCommentFmt(L"Preferred shader wave size: %d", prefShaderWaveSize);
+
+        // at this point we assume that the waverange size that
+        // the shader specifies is legal.
+        test->Test->GetReadBackData("UAVBuffer0", &dataUav);
+        VERIFY_ARE_EQUAL(sizeof(WaveSizeTestData) * MAX_WAVESIZE,
+                         dataUav.size());
+        pOutData = (WaveSizeTestData *)dataUav.data();
+
+        for (unsigned i = 0; i < MAX_WAVESIZE; i++) {
+          VERIFY_IS_TRUE(pOutData[i].count >= minWaveSize &&
+                         pOutData[i].count <= maxWaveSize);
+
+          if (prefShaderWaveSize >= minWaveSize &&
+              prefShaderWaveSize <= maxWaveSize) {
+
+            VERIFY_ARE_EQUAL(pOutData[i].count, prefShaderWaveSize);
+          }
```

Comment:
Seems like the first VERIFY should be in an else clause when not checking `prefShaderWaveSize`.
Also, we should use the appropriate comparison VERIFY macros so you see value being compared when it fails.

```suggestion
          unsigned count = pOutData[i].count;
          if (prefShaderWaveSize >= minWaveSize &&
              prefShaderWaveSize <= maxWaveSize) {
            VERIFY_ARE_EQUAL(count, prefShaderWaveSize);
          } else {
            VERIFY_IS_GREATER_THAN_OR_EQUAL(count, minWaveSize);
            VERIFY_IS_LESS_THAN_OR_EQUAL(count, maxWaveSize);
          }
```